### PR TITLE
Spoiler flags

### PR DESCRIPTION
**Pull request by change1331**
Adds start, ship and flags to spoiler map.  Detailed flags in spoiler sidebar.  Step one displayed by default if available. Clicking start location shows step one area and route to hub. Clicking ship gives escape route.  Settings for everything on the map.  Help gets displayed by default with button to not show again.

### CODE REVIEW

**Comment by blkerby** [comment]
I notice that clicking on "relevant flags" links inside of route details isn't working; it gives this error:

```
Uncaught TypeError: k.relevant_flags is undefined
    onclick http://localhost:8080/seed/vq6Sjxt7r/data/visualizer/script.js:422
    routeData http://localhost:8080/seed/vq6Sjxt7r/data/visualizer/script.js:421
    show_item_details http://localhost:8080/seed/vq6Sjxt7r/data/visualizer/script.js:351
    onclick http://localhost:8080/seed/vq6Sjxt7r/data/visualizer/script.js:730
[script.js:422:27](http://localhost:8080/seed/vq6Sjxt7r/data/visualizer/script.js)
```

If I understand, it looks like the issue is `k` needs to be declared as `var k`, otherwise it's treated as global and won't get captured in the closure that's assigned to `flagA.onclick`.

**Comment by blkerby** [comment]
Great stuff, it's really coming together!

Couple other things I noticed:
- The item in Speedbooster Hall is showing at the top-left instead of bottom-right.
- I think a flag shouldn't be drawn on the map for Bomb Torizo. It doesn't fit with the item there; shifting the item over has an issue that it looks wrong if flags are disabled from being drawn. The logic for Bomb Torizo is not very important anyway since it is so easy, and the player can find it from the side panel if they want.

**Comment by change1331** [comment]
Fixed up speedbooster it was a room not a hall.  Good catch on the relevant flags links.

File: visualizer/index.html [hunk 1]
```diff
@@ -1,39 +1,77 @@
...
+                <li>Select a step in the overview (top-left corner) to focus on logically reachable areas for that step.
+                <li>One-way-reachable areas appear darker.
+                <li>Only the first instance of each item is shown in the overview.
+                <li>Double-click a step to show details, including all items collectible on that step.
+                <li>Click an item to show a path to reach it (white) and return to the Ship (yellow).
+                <li>Double-click a room to pull up the logic for that room.
+                <li>Click on a room containing a boss, objective, or other flag to show logic for completing it.
+                <li>Click on the Mother Brain Room to show the intended escape route.
+                <li>Where an item routes shows a difficulty level, it represents only the difficulty of reaching the item;
 					it does not include the difficulty of returning, nor of any flags assumed to be previously cleared (e.g. bosses).
```
Comment (kjbranch) [inline visualizer/index.html:24]:
```suggestion
                <li>Click on the Mother Brain Room or the Ship icon to show the intended escape route.
                <li>Where an item route shows a difficulty level, it represents only the difficulty of reaching the item;
					it does not include the difficulty of returning, nor of any flags cleared (e.g. bosses).
```

**Review by kjbranch — COMMENTED** [review]
This is looking really nice and seems close to being finished; just a couple more suggestions:
- [x] I think the BT icon should be removed (or maybe it could look okay if it were on the left and the item on the right, so if it is disabled, the door will be visible?)
- [ ] I see that the last step is not openable by double clicking anymore (it may only be a problem if there are no key items on the last step?). I see clicking on something obtainable in the last step (MB) can still open it.it, or maybe there's something more subtle going on that isn't related to this PR
- [x] Icons on the map are a bit blurry, is there a way to fix that? (Item icons are not blurry)
- [x] In game, the ship icon is 2x1 map tiles. I think here it could be expanded by ~1.5x to look right
- [x] The green circle around the objective flags seems unnecessary since there are the visible X's behind it. Those circles can be removed.
- [x] I'm thinking the flag could be a bit more visible with a darker border, so I can send you a new icon for it.

**Comment by kjbranch** [comment]
Generated a new seed with your recent commit. The last step is now openable, so maybe you fixed it, or maybe there's something more subtle going on that isn't related to this PR

I notice that the difficulty of each step is also not there, so I'm wondering if an error in this PR could have cause that (the other seed I generated also didnt have them, but it was `Normal`, so I dont think it usually does)

Unchecking `Objectives` to remove the green circle is much better, but I still think it would be best to just remove the circle altogether.

**Comment by blkerby** [comment]
> Unchecking `Objectives` to remove the green circle is much better, but I still think it would be best to just remove the circle altogether.

Yeah I agree the circles seem unnecessary. We already have X map tiles to mark the objectives, and the purpose of making the icons smaller was to avoid obscuring the map tiles as much; the circles kind of run counter to that goal. If the circles are removed, then I assume we'd also remove the "objectives" checkbox under the settings cog. It's kind of awkward how it overlaps with the other categories, so removing it will clean things up I think.

A few other thoughts:

- The flag icons are getting nice enough to where we could leave them on by default. And in this case, I think it makes sense to remove the old behavior where clicking certain rooms themselves would open a route; that's no longer necessary since now it's possible to click on the flag icons. And even if players disable the flag icons, there are now more ways to open them using the side panel so I think it will be fine.

- One thing I'd like to see is if clicking on the start location would pull up the step-by-step route in the side panel, similar to how it does when clicking on an item or flag icon. Currently it draws the path on the map, but this won't always give the details that the player may be looking for.

- The remaining "flag" categories "start", "ship", "bosses", "minibosses", and "misc" all have distinct icons except that bosses and minibosses use the same icon. If we can think of a different icon for minibosses, that could be nice to iron out that inconsistency and can make the symbols a little easier for the user to parse. Another alternative would be to lump "minibosses" into the same "bosses" category instead of having a separate checkbox for it. Not sure what is best, just throwing out ideas

- When hovering over the hub room, I think I'd prefer if it didn't use a second line of text just for the word "Hub". It could be written in parentheses on the same line, e.g. "Kraid Eye Door Room (Hub)".

**Comment by kjbranch** [comment]
> * The remaining "flag" categories "start", "ship", "bosses", "minibosses", and "misc" all have distinct icons except that bosses and minibosses use the same icon. If we can think of a different icon for minibosses, that could be nice to iron out that inconsistency and can make the symbols a little easier for the user to parse. Another alternative would be to lump "minibosses" into the same "bosses" category instead of having a separate checkbox for it. Not sure what is best, just throwing out ideas

I agree, but couldnt think of any icon to use. Maybe a recolor on the boss icon? I can look into some ideas

**Comment by kjbranch** [comment]
> * The flag icons are getting nice enough to where we could leave them on by default. And in this case, I think it makes sense to remove the old behavior where clicking certain rooms themselves would open a route; that's no longer necessary since now it's possible to click on the flag icons. And even if players disable the flag icons, there are now more ways to open them using the side panel so I think it will be fine.

Maybe the optimal default would be bosses/start/ship on, minis/misc off? 
That's very close to how it would look in game if we were to put icons in game (just adding the start icon). I'm also not overly loving the flag icon and could see it adding clutter, and I'm not confident in my abilities to make a miniboss icon.

**Comment by blkerby** [comment]
I think if any of these are default-on, minibosses should be included: Crocomire is the biggest one that I see creating trouble for people, where it would be helpful to be able to click on it to see the route. I think I'm liking the idea more of simplifying the options: "items" could all be collapsed to a single checkbox (no real reason to want to toggle individual categories), "minibosses" could be merged with "bosses". With "objectives" removed, "misc" could just be renamed to "flags". Then yeah, maybe "bosses" could be default-on while "flags" could be default-off, since it is more rare for those other flags to be important in the logic, though I don't think they necessarily look bad either.

**Comment by kjbranch** [comment]
Talking with Maddo a bit, we were thinking of flattening the checkboxes, like so:

- [ ] start
- [ ] ship
- [ ] items
- [ ] bosses
- [ ] minibosses
- [ ] flags
- [ ] objectives

We aren't sure if we want to be able to toggle different types of items at this time, partly because it makes some of them look odd when one of the two icons on a tile is hidden. Maybe you could keep the code in there separating them, but just collapse it to a single checkbox, so if we change our mind in the future?

I was then thinking that bosses/minibosses/flags could be separated into different icons: a green boss icon for minibosses, and a gold one for either bosses/minibosses if the objectives checkbox is selected (otherwise it would only use the green/grey variants). Similarly, a gold flag could be used for objective flags, if the objectives checkbox is selected. I'll send you these new icons - what do you think of them and this idea?

**Comment by change1331** [comment]
Actually trying flattening right now went all the way down to start/ship/items/bosses/misc with minibosses in bosses.  Was mildy considering adding in dynamic offsets for bomb torizo could probably do same for items in same room already have it for start loc + item(s).

**Comment by kjbranch** [comment]
I think I'm satisfied with everything except how the objectives checkbox works. When unchecking it,  but keeping everything else on, I was hoping that all the icons would be on the map, using the miniboss/boss/white flag icons. When having it checked then it would convert the 4 objectives + MB to use the gold flag and gold skull icons. 

Currently, unchecking it removes the icons from the objectives, but if you toggle on/off other checkboxes, some of the objective icons will return, but will be their gold color. Also when doing a seed with "Bosses" objectives and unchecking objectives, it unchecks the "bosses" checkbox.

I do still wonder if it would make more sense to name `misc` as `flags`, for clarity.

**Comment by kjbranch** [comment]
We are thinking of having the objectives checkbox off by default, where the rest will be on by default. If fixing the objectives checkbox is going to be a lot of work, it may just be worth not having a checkbox for it.

What are your thoughts? Do you think the gold icons look nice? Should the gold skull be changed into a red one?

**Comment by change1331** [comment]
Easy to change.  Gold or red is good used same boss obj for minibosses tried bright green but thought it should be orange to keep consistancy worth testing other colors to see what stands out for objectives.  Flags instead of misc is good idea as they are literal flags now.

**Review by kjbranch — APPROVED** [review]
Everything looks good to me.

**Pull request merged**